# Initial Cleaning of Students' Assessment Data

This exercise used two data sets from the Open University Learning Analytics (OULAD) project to illustrate initial cleaning. According to The Open University web site, the data is about "courses, students and their interactions with Virtual Learning Environment (VLE) for seven selected courses (called modules)".

Initial cleaning activities such as "humanizing" field names, changing data types, and removing unnecessary data are illustrated. Additionally, merging datasets is illustrated. This would be the first part of a process ultimately leading to further exploration, visualization, and statistical and predictive analysis based on specific business requirements.

This the [Open Learning Analytics dataset](https://analyse.kmi.open.ac.uk/open_dataset) used in this project

In [1]:
import pandas as pd
import numpy as np
assessments = pd.read_csv("C:\dataHub\studentAssessment.csv")
students_demo = pd.read_csv("C:\dataHub\studentInfo.csv") # default encoding

In [2]:
assessments.info() # view datatypes of the first dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
id_assessment     173912 non-null int64
id_student        173912 non-null int64
date_submitted    173912 non-null int64
is_banked         173912 non-null int64
score             173912 non-null int64
dtypes: int64(5)
memory usage: 6.6 MB


In [3]:
assessments.sample(n = 3) # spot check the first dataset

id_assessment  id_student  date_submitted  is_banked  score
63178           34880      501871             227          0     73
157349          15007      610362             210          0    100
28387           37429      607762             222          0     60

In [4]:
students_demo.info() # view data types of the second dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
code_module             32593 non-null object
code_presentation       32593 non-null object
id_student              32593 non-null int64
gender                  32593 non-null object
region                  32593 non-null object
highest_education       32593 non-null object
imd_band                31482 non-null object
age_band                32593 non-null object
num_of_prev_attempts    32593 non-null int64
studied_credits         32593 non-null int64
disability              32593 non-null object
final_result            32593 non-null object
dtypes: int64(3), object(9)
memory usage: 3.0+ MB


In [5]:
students_demo.sample(n = 3) # spot check the second dataset

code_module code_presentation  id_student gender                region  \
29522         FFF             2014J      685212      M  West Midlands Region   
22618         FFF             2013B      393934      M   East Anglian Region   
30449         GGG             2013J      585025      M          South Region   

           highest_education imd_band age_band  num_of_prev_attempts  \
29522       HE Qualification   30-40%    35-55                     0   
22618  A Level or Equivalent   30-40%     0-35                     0   
30449  A Level or Equivalent    10-20     0-35                     0   

       studied_credits disability final_result  
29522              120          N  Distinction  
22618              120          N         Fail  
30449               30          N         Fail

Assessments dataset contains 173,912 rows and 5 columns, while the student demographics dataset consists of 32,593 entries and 12 columns. 

Both datasets contain data types and columns that requires cleaning and preparation for analysis, so we shall begin with that.

# Data Preparation

### Rename Columns

In [6]:
assessments.columns

Index(['id_assessment', 'id_student', 'date_submitted', 'is_banked', 'score'], dtype='object')

In [7]:
# rename 'date_submitted', 'is_banked', and 'score' columns
assessments.columns = ['id_assessment', 'id_student', 'time_taken_days', 'result_transferred_from_previous', 'score_pct']

In [8]:
assessments.columns

Index(['id_assessment', 'id_student', 'time_taken_days',
       'result_transferred_from_previous', 'score_pct'],
      dtype='object')

In [9]:
assessments.sample(n = 3) # spot check first dataset after column renaming

id_assessment  id_student  time_taken_days  \
56862           24286      620369               20   
169886          15003      563706               56   
152379          34907      646074              233   

        result_transferred_from_previous  score_pct  
56862                                  0         70  
169886                                 0        100  
152379                                 0         98

In [10]:
students_demo.columns

Index(['code_module', 'code_presentation', 'id_student', 'gender', 'region',
       'highest_education', 'imd_band', 'age_band', 'num_of_prev_attempts',
       'studied_credits', 'disability', 'final_result'],
      dtype='object')

In [11]:
# rename 'highest_edu', 'num_of_prev_attempts', 'studied_credits', 'imd_band', and 'disability' columns
students_demo.columns = ['code_module', 'code_presentation', 'id_student', 'gender', 'region',
       'highest_edu', 'imd_band_pct', 'age_band', 'num_prev_attempts', 'current_credits', 'disabled', 'final_result']

In [12]:
students_demo.columns

Index(['code_module', 'code_presentation', 'id_student', 'gender', 'region',
       'highest_edu', 'imd_band_pct', 'age_band', 'num_prev_attempts',
       'current_credits', 'disabled', 'final_result'],
      dtype='object')

In [13]:
students_demo.sample(n = 3) # spot check second dataset after column renaming

code_module code_presentation  id_student gender                region  \
26790         FFF             2014B      596699      M   East Anglian Region   
10123         CCC             2014B      627176      F      Yorkshire Region   
15628         DDD             2013J      585929      M  North Western Region   

                 highest_edu imd_band_pct age_band  num_prev_attempts  \
26790       HE Qualification       30-40%     0-35                  1   
10123     Lower Than A Level       50-60%     0-35                  0   
15628  A Level or Equivalent        10-20     0-35                  0   

       current_credits disabled final_result  
26790              240        N         Fail  
10123               90        N    Withdrawn  
15628              120        N    Withdrawn

### Explore and Clean Data

#### Assessments Dataset

In [14]:
# start exploring assessments data
assessments.describe(include = 'all')

id_assessment    id_student  time_taken_days  \
count  173912.000000  1.739120e+05    173912.000000   
mean    26553.803556  7.051507e+05       116.032942   
std      8829.784254  5.523952e+05        71.484148   
min      1752.000000  6.516000e+03       -11.000000   
25%     15022.000000  5.044290e+05        51.000000   
50%     25359.000000  5.852080e+05       116.000000   
75%     34883.000000  6.344980e+05       173.000000   
max     37443.000000  2.698588e+06       608.000000   

       result_transferred_from_previous      score_pct  
count                     173912.000000  173912.000000  
mean                           0.010977      75.724171  
std                            0.104194      18.940093  
min                            0.000000       0.000000  
25%                            0.000000      65.000000  
50%                            0.000000      80.000000  
75%                            0.000000      90.000000  
max                            1.000000     100.000000

All data are currently integer types. To perform useful statistical or mathematical calculations, only 'time_taken_days' and 'score_pct' data is required as integers from this dataset. 
To avoid clutter, we can change the rest of the data to text.
Additionally, the 'result_transferred_from_previous' column needs to be double-checked based on initial exploration.

In [15]:
assessments['result_transferred_from_previous'].value_counts() # it's binary data (Y/N) so that is OK

0    172003
1      1909
Name: result_transferred_from_previous, dtype: int64

In [16]:
# convert some column types to text and replace zeros and ones with No and Yes
assessments['id_assessment'] = assessments['id_assessment'].astype(str)
assessments['id_student'] = assessments['id_student'].astype(str)
assessments['result_transferred_from_previous'] = (assessments['result_transferred_from_previous']
                                                   .astype(str)
                                                   .str.replace("0", "No")
                                                   .str.replace("1", "Yes")
                                                  )

In [17]:
assessments.describe(include = 'all')

id_assessment id_student  time_taken_days  \
count         173912     173912    173912.000000   
unique           188      23369              NaN   
top            24295     537811              NaN   
freq            1917         28              NaN   
mean             NaN        NaN       116.032942   
std              NaN        NaN        71.484148   
min              NaN        NaN       -11.000000   
25%              NaN        NaN        51.000000   
50%              NaN        NaN       116.000000   
75%              NaN        NaN       173.000000   
max              NaN        NaN       608.000000   

       result_transferred_from_previous      score_pct  
count                            173912  173912.000000  
unique                                2            NaN  
top                                  No            NaN  
freq                             172003            NaN  
mean                                NaN      75.724171  
std                                 NaN      18.940093  
min                                 NaN       0.000000  
25%                                 NaN      65.000000  
50%                                 NaN      80.000000  
75%                                 NaN      90.000000  
max                                 NaN     100.000000

In [18]:
assessments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
id_assessment                       173912 non-null object
id_student                          173912 non-null object
time_taken_days                     173912 non-null int64
result_transferred_from_previous    173912 non-null object
score_pct                           173912 non-null int64
dtypes: int64(2), object(3)
memory usage: 6.6+ MB


In [19]:
# spot check that the previous zeros and ones correspond to the No and Yes
assessments['result_transferred_from_previous'].value_counts()

No     172003
Yes      1909
Name: result_transferred_from_previous, dtype: int64

In [20]:
# final spot check
assessments.sample(n=5)

id_assessment id_student  time_taken_days  \
95393          15003     138920               56   
46933          25365     684936              111   
47463          25356     634246               47   
154796         30719     344740               31   
22344          24299    1091274              244   

       result_transferred_from_previous  score_pct  
95393                                No         80  
46933                                No         67  
47463                                No         67  
154796                               No         99  
22344                                No         56

The assessments dataset is now clean and ready to be joined to the student information dataset. First, however, we need to ensure the student information dataset is also clean

#### Student Demographics Dataset

In [21]:
# start exploring students' demographics data
students_demo.describe(include = 'all')

code_module code_presentation    id_student gender    region  \
count        32593             32593  3.259300e+04  32593     32593   
unique           7                 4           NaN      2        13   
top            BBB             2014J           NaN      M  Scotland   
freq          7909             11260           NaN  17875      3446   
mean           NaN               NaN  7.066877e+05    NaN       NaN   
std            NaN               NaN  5.491673e+05    NaN       NaN   
min            NaN               NaN  3.733000e+03    NaN       NaN   
25%            NaN               NaN  5.085730e+05    NaN       NaN   
50%            NaN               NaN  5.903100e+05    NaN       NaN   
75%            NaN               NaN  6.444530e+05    NaN       NaN   
max            NaN               NaN  2.716795e+06    NaN       NaN   

                  highest_edu imd_band_pct age_band  num_prev_attempts  \
count                   32593        31482    32593       32593.000000   
unique                      5           10        3                NaN   
top     A Level or Equivalent       20-30%     0-35                NaN   
freq                    14045         3654    22944                NaN   
mean                      NaN          NaN      NaN           0.163225   
std                       NaN          NaN      NaN           0.479758   
min                       NaN          NaN      NaN           0.000000   
25%                       NaN          NaN      NaN           0.000000   
50%                       NaN          NaN      NaN           0.000000   
75%                       NaN          NaN      NaN           0.000000   
max                       NaN          NaN      NaN           6.000000   

        current_credits disabled final_result  
count      32593.000000    32593        32593  
unique              NaN        2            4  
top                 NaN        N         Pass  
freq                NaN    29429        12361  
mean          79.758691      NaN          NaN  
std           41.071900      NaN          NaN  
min           30.000000      NaN          NaN  
25%           60.000000      NaN          NaN  
50%           60.000000      NaN          NaN  
75%          120.000000      NaN          NaN  
max          655.000000      NaN          NaN

In [22]:
students_demo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
code_module          32593 non-null object
code_presentation    32593 non-null object
id_student           32593 non-null int64
gender               32593 non-null object
region               32593 non-null object
highest_edu          32593 non-null object
imd_band_pct         31482 non-null object
age_band             32593 non-null object
num_prev_attempts    32593 non-null int64
current_credits      32593 non-null int64
disabled             32593 non-null object
final_result         32593 non-null object
dtypes: int64(3), object(9)
memory usage: 3.0+ MB


Students' demographics data types are currently a mix of text and integers. Here, we shall check and clean data and data types where appropriate. For example, student ids do not need to be integers.

In [23]:
# convert students' id data type from integer to string
students_demo['id_student'] = students_demo['id_student'].astype(str)

In [24]:
students_demo.info() # confirm the conversion

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 12 columns):
code_module          32593 non-null object
code_presentation    32593 non-null object
id_student           32593 non-null object
gender               32593 non-null object
region               32593 non-null object
highest_edu          32593 non-null object
imd_band_pct         31482 non-null object
age_band             32593 non-null object
num_prev_attempts    32593 non-null int64
current_credits      32593 non-null int64
disabled             32593 non-null object
final_result         32593 non-null object
dtypes: int64(2), object(10)
memory usage: 3.0+ MB


Given that the imd_band_pct column contains percentage (%) text that potentially hampers analysis, it is best to remove the '%' part of the band

In [29]:
students_demo["imd_band_pct"] = students_demo["imd_band_pct"].str.replace("%", "")

In [30]:
students_demo["imd_band_pct"].sample(n = 3) # confirm removal of '%'

27629    60-70
7220      0-10
6312     10-20
Name: imd_band_pct, dtype: object

In [31]:
students_demo.sample(n = 5)

code_module code_presentation id_student gender                region  \
19383         EEE             2013J     113824      M   East Anglian Region   
25215         FFF             2013J     585731      M              Scotland   
8517          BBB             2014J    1929491      F  East Midlands Region   
29154         FFF             2014J     656734      F     South West Region   
28760         FFF             2014J     645880      M              Scotland   

                 highest_edu imd_band_pct age_band  num_prev_attempts  \
19383     Lower Than A Level       90-100     0-35                  0   
25215  A Level or Equivalent         0-10     0-35                  0   
8517        HE Qualification        20-30    35-55                  0   
29154     Lower Than A Level        70-80     0-35                  0   
28760  A Level or Equivalent         0-10     0-35                  0   

       current_credits disabled final_result  
19383              120        Y         Fail  
25215               60        N         Pass  
8517                60        N    Withdrawn  
29154              120        N  Distinction  
28760               60        N         Fail

The students demographic data has been cleaned where appropriate.
The next step is to merge the assessments and demographics datasets. We shall use the 'id_student' field to merge the two datasets.

In [37]:
combined = assessments.merge(students_demo, on=['id_student']) # merge using students' IDs
combined.dropna(inplace = True)

In [38]:
combined.describe(include="all") # check the newly merged dataset

id_assessment id_student  time_taken_days  \
count         198004     198004    198004.000000   
unique           188      22488              NaN   
top            24295     610287              NaN   
freq            2876         92              NaN   
mean             NaN        NaN       114.263924   
std              NaN        NaN        72.332765   
min              NaN        NaN       -11.000000   
25%              NaN        NaN        49.000000   
50%              NaN        NaN       114.000000   
75%              NaN        NaN       172.000000   
max              NaN        NaN       608.000000   

       result_transferred_from_previous      score_pct code_module  \
count                            198004  198004.000000      198004   
unique                                2            NaN           7   
top                                  No            NaN         FFF   
freq                             194590            NaN       56850   
mean                                NaN      75.142997         NaN   
std                                 NaN      19.295583         NaN   
min                                 NaN       0.000000         NaN   
25%                                 NaN      65.000000         NaN   
50%                                 NaN      79.000000         NaN   
75%                                 NaN      89.000000         NaN   
max                                 NaN     100.000000         NaN   

       code_presentation  gender    region            highest_edu  \
count             198004  198004    198004                 198004   
unique                 4       2        13                      5   
top                2014J       M  Scotland  A Level or Equivalent   
freq               61773  111905     22627                  92921   
mean                 NaN     NaN       NaN                    NaN   
std                  NaN     NaN       NaN                    NaN   
min                  NaN     NaN       NaN                    NaN   
25%                  NaN     NaN       NaN                    NaN   
50%                  NaN     NaN       NaN                    NaN   
75%                  NaN     NaN       NaN                    NaN   
max                  NaN     NaN       NaN                    NaN   

       imd_band_pct age_band  num_prev_attempts  current_credits disabled  \
count        198004   198004      198004.000000    198004.000000   198004   
unique           10        3                NaN              NaN        2   
top           30-40     0-35                NaN              NaN        N   
freq          22785   137662                NaN              NaN   179308   
mean            NaN      NaN           0.158360        78.118902      NaN   
std             NaN      NaN           0.464224        38.025589      NaN   
min             NaN      NaN           0.000000        30.000000      NaN   
25%             NaN      NaN           0.000000        60.000000      NaN   
50%             NaN      NaN           0.000000        60.000000      NaN   
75%             NaN      NaN           0.000000        90.000000      NaN   
max             NaN      NaN           6.000000       630.000000      NaN   

       final_result  
count        198004  
unique            4  
top            Pass  
freq         112509  
mean            NaN  
std             NaN  
min             NaN  
25%             NaN  
50%             NaN  
75%             NaN  
max             NaN

Immediately after merging, it is a good idea to cross-check the combined dataset with the original datasets

In [40]:
combined.shape # the combined dataset's structure.
               # there are now 16 columns (one less after merging 'id_student')

(198004, 16)

In [43]:
combined.sample(n = 5)

id_assessment id_student  time_taken_days  \
146603         15003     318848               56   
79254          24296     515857               69   
84145          25350     361243               87   
68202          24291     588493               32   
102761         24290     543170              230   

       result_transferred_from_previous  score_pct code_module  \
146603                               No         80         BBB   
79254                                No         74         FFF   
84145                                No         81         DDD   
68202                                No         76         CCC   
102761                               No         72         CCC   

       code_presentation gender                region         highest_edu  \
146603             2013J      F     South East Region  Lower Than A Level   
79254              2013J      M  East Midlands Region  Lower Than A Level   
84145              2013J      F          South Region  Lower Than A Level   
68202              2014J      M              Scotland    HE Qualification   
102761             2014B      M  North Western Region  Lower Than A Level   

       imd_band_pct age_band  num_prev_attempts  current_credits disabled  \
146603        70-80    35-55                  0               60        N   
79254         10-20    35-55                  0               60        N   
84145         70-80     0-35                  0               90        N   
68202        90-100     0-35                  0               60        N   
102761         0-10     0-35                  0               90        Y   

       final_result  
146603         Pass  
79254          Pass  
84145          Pass  
68202          Fail  
102761         Pass

## Conclusion

Two datasets, one for students assessments and a second for demographics, were explored and updated to ensure seamless future analysis. The preparation included renaming columns to make them more meaningful, updating data types based on intended use, and cleaning out unnecessary data such as percentage signs. 

A next step would be to explore the combined dataset by looking at each series more closely, aggregating, and visualizing the most pertinent aspects based on business requirements.